Use `ThinVec` more in the AST

nnethercote opened the pull request:
r? @ghost

nnethercote committed:
Use `ThinVec` in `ast::Generics` and related types.

nnethercote committed:
Use `ThinVec` in `ast::WhereClause`.

nnethercote committed:
Use `ThinVec` in `ast::Impl` and related types.

nnethercote committed:
Use `ThinVec` in various AST types.

This commit changes the sequence parsers to produce `ThinVec`, which
triggers numerous conversions.

nnethercote commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit 26324d59c93e8eb9d08792a57b02990185a45852 with merge 10ea77cebafccd3c9085348ddf5b678008e8a87c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"26324d59c93e8eb9d08792a57b02990185a45852","merge_sha":"10ea77cebafccd3c9085348ddf5b678008e8a87c"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3528584386/jobs/5919803792)
Build commit: 10ea77cebafccd3c9085348ddf5b678008e8a87c (`10ea77cebafccd3c9085348ddf5b678008e8a87c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3528584386/jobs/5919803792"},"merge_sha":"10ea77cebafccd3c9085348ddf5b678008e8a87c"} -->

rust-timer commented:
Queued 10ea77cebafccd3c9085348ddf5b678008e8a87c with parent ff8c8dfbe66701531e3e5e335c28c544d0fbc945, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ff8c8dfbe66701531e3e5e335c28c544d0fbc945&end=10ea77cebafccd3c9085348ddf5b678008e8a87c).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (10ea77cebafccd3c9085348ddf5b678008e8a87c): [comparison URL](https://perf.rust-lang.org/compare.html?start=ff8c8dfbe66701531e3e5e335c28c544d0fbc945&end=10ea77cebafccd3c9085348ddf5b678008e8a87c&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ff8c8dfbe66701531e3e5e335c28c544d0fbc945&end=10ea77cebafccd3c9085348ddf5b678008e8a87c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.3%, 1.2%]   | 14    |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.9%, -0.2%] | 29    |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.8%, -0.1%] | 20    |
| All ❌✅ (primary)                 | -0.4% | [-0.9%, 0.4%]  | 32    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ff8c8dfbe66701531e3e5e335c28c544d0fbc945&end=10ea77cebafccd3c9085348ddf5b678008e8a87c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.5%  | [1.2%, 3.6%]   | 16    |
| Improvements ✅ <br /> (primary)   | -1.3% | [-2.1%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-3.7%, -1.2%] | 3     |
| All ❌✅ (primary)                 | -1.3% | [-2.1%, -0.5%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ff8c8dfbe66701531e3e5e335c28c544d0fbc945&end=10ea77cebafccd3c9085348ddf5b678008e8a87c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.4% | [-4.4%, -4.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

nnethercote commented:
Let's try again, see if the bootstrapping improvement persists.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 4ce53b1508cf82969ed83a009865aaece57280e8 with merge e027f83413d67db20d22ecd9ad55315b279b09e2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4ce53b1508cf82969ed83a009865aaece57280e8","merge_sha":"e027f83413d67db20d22ecd9ad55315b279b09e2"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3544272514/jobs/5952310595)
Build commit: e027f83413d67db20d22ecd9ad55315b279b09e2 (`e027f83413d67db20d22ecd9ad55315b279b09e2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3544272514/jobs/5952310595"},"merge_sha":"e027f83413d67db20d22ecd9ad55315b279b09e2"} -->

rust-timer commented:
Queued e027f83413d67db20d22ecd9ad55315b279b09e2 with parent b3bc6bf31265ac10946a0832092dbcedf9b26805, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b3bc6bf31265ac10946a0832092dbcedf9b26805&end=e027f83413d67db20d22ecd9ad55315b279b09e2).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e027f83413d67db20d22ecd9ad55315b279b09e2): [comparison URL](https://perf.rust-lang.org/compare.html?start=b3bc6bf31265ac10946a0832092dbcedf9b26805&end=e027f83413d67db20d22ecd9ad55315b279b09e2&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=b3bc6bf31265ac10946a0832092dbcedf9b26805&end=e027f83413d67db20d22ecd9ad55315b279b09e2&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.5%, 1.2%]   | 5     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.3%, -0.2%] | 77    |
| Improvements ✅ <br /> (secondary) | -0.6% | [-1.3%, -0.2%] | 38    |
| All ❌✅ (primary)                 | -0.5% | [-1.3%, -0.2%] | 77    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b3bc6bf31265ac10946a0832092dbcedf9b26805&end=e027f83413d67db20d22ecd9ad55315b279b09e2&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.2%  | [1.8%, 6.8%]   | 18    |
| Improvements ✅ <br /> (primary)   | -1.3% | [-2.9%, -0.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-3.2%, -2.1%] | 2     |
| All ❌✅ (primary)                 | -1.3% | [-2.9%, -0.5%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b3bc6bf31265ac10946a0832092dbcedf9b26805&end=e027f83413d67db20d22ecd9ad55315b279b09e2&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [2.3%, 2.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

nnethercote commented:
Wow, this time the bootstrap improvement was 7.9 seconds (1.15%). And the instruction count results were good too. I think  this is worth going forward.

r? @the8472

the8472 commented on src/tools/clippy/clippy_lints/src/unnested_or_patterns.rs:
```diff
@@ -178,7 +178,12 @@ fn unnest_or_patterns(pat: &mut P<Pat>) -> bool {
                 this_level_changed = true;
-                alternatives.splice(idx..=idx, inner);
+ 
+                // Equivalent to `alternatives.splice(idx..=idx, inner);`.
+                alternatives.remove(idx);
+                for p in inner.into_iter().rev() {
+                    alternatives.insert(idx, p);
+                }
```
This looks algorithmically worse since it'll move the tail on every insert. The splice impl moves the tail by the size hint, inserts the elements, then moves it back.

I have no idea how hot this code is of course. Does clippy have some benchmarks?

nnethercote replied:
Not that I know of. `ThinVec` doesn't have `splice`, alas.

the8472 replied:
So I assume this case will be uncommon to hit, especially uncommon with large enough N that it matters. But I still am concerned about letting an N² footgun lie around. At least a FIXME, an issue to get splice into thinvec or something would be good.

Or if thin-vec had cheap conversions between itself and vec then maybe that could be used.

With that addressed r=me.

nnethercote replied:
I filed https://github.com/Gankra/thin-vec/issues/39 about adding `ThinVec::splice` and added a FIXME comment that refers to it.

Cheap conversions between `ThinVec` and `Vec` aren't possible because the memory layout is different, alas.

nnethercote replied:
Good news: `thin-vec` 0.2.10 was just released and it includes support for `splice`. Thanks, @Gankra! So I have changed this back to the original code.

the8472 commented on src/tools/clippy/clippy_lints/src/unnested_or_patterns.rs:
```diff
@@ -392,26 +397,30 @@ fn extend_with_tail_or(target: &mut Pat, tail_or: Vec<P<Pat>>) -> bool {
-        tail_or.push(extract(pat.into_inner().kind));
+        // Check if we should extract, but only if `idx >= start`.
+	if idx > start && predicate(&alternatives[i].kind) {
+	    let pat = alternatives.remove(i);
```
This could also be algorithmically worse than drain_filter since that one avoids moves for runs of removed elements. But that's data-dependent, the worst case isn't getting worse.

nnethercote replied:
Yep. `ThinVec` also lacks `drain_filter`.

the8472 commented on compiler/rustc_ast/src/ast.rs:
```diff
@@ -1339,11 +1339,11 @@ pub enum ExprKind {
     MethodCall(Box<MethodCall>),
     /// A tuple (e.g., `(a, b, c, d)`).
-    Tup(Vec<P<Expr>>),
+    Tup(ThinVec<P<Expr>>),
```
Possibly stupid questions ahead, I don't know much about the AST.
Unit types count as tuple I assume. Would having a separate `Unit()` variant make a difference here? Then the other ones would always be non-empty.

nnethercote replied:
There would be no space/speed benefit to having a separate `Unit`. The `Tup` case would still be the same size. In fact, an empty `ThinVec` is the special case -- it's just a single pointer to a static thing -- so there's no point trying to handle it separately.

the8472 commented on src/test/ui/stats/hir-stats.stderr:
```diff
@@ -1,119 +1,119 @@
+ast-stats-2 - Use                      576 ( 7.7%)             4
 ast-stats-2 ----------------------------------------------------------------
-ast-stats-2 Total                  8_112
+ast-stats-2 Total                  7_464
```
🎉

the8472 reviewed (commented):
Overall a fairly mechanical replacement that looks fine. Only two issues in clippy code. Those probably would only become an issue if someone code-generates huge patterns and then runs clippy on them

nnethercote commented:
I rebased.

bors commented:
:umbrella: The latest upstream changes (presumably #104861) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I addressed the comments and rebased.

@bors r=the8472

bors commented:
:pushpin: Commit 09d934060e91a1f202a87f4683fa0110ae674d86 has been approved by `the8472`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=the8472 09d934060e91a1f202a87f4683fa0110ae674d86 -->
<!-- homu: {"type":"Approved","sha":"09d934060e91a1f202a87f4683fa0110ae674d86","approver":"the8472","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
Whoops, I forgot to commit the final changes before pushing.

@bors r=the8472

bors commented:
:pushpin: Commit e02b396d71d26f8899f797c49ee9c57254de1b48 has been approved by `the8472`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=the8472 e02b396d71d26f8899f797c49ee9c57254de1b48 -->
<!-- homu: {"type":"Approved","sha":"e02b396d71d26f8899f797c49ee9c57254de1b48","approver":"the8472","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e02b396d71d26f8899f797c49ee9c57254de1b48 with merge fc3a8ed02f357f23795ee7e8d38cf275f1b2a950...
<!-- homu: {"type":"BuildStarted","head_sha":"e02b396d71d26f8899f797c49ee9c57254de1b48","merge_sha":"fc3a8ed02f357f23795ee7e8d38cf275f1b2a950"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3597630449/jobs/6062032355) [(plain)](https://github.com/rust-lang-ci/rust/commit/fc3a8ed02f357f23795ee7e8d38cf275f1b2a950/checks/9839181294/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

nnethercote commented:
It appears the `dist-x86_64-linux-alt, dist-x86_64-linux, ubuntu-20.04-xl` job timed out? It ran for 4h17m and ended with:
```
[RUSTC-TIMING] rustc_privacy test:false 0.506
 Documenting rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
[RUSTC-TIMING] rustc_borrowck test:false 5.778
[RUSTC-TIMING] rustc_codegen_llvm test:false 3.473
[RUSTC-TIMING] rustc_hir_typeck test:false 7.157
    Checking rustc_interface v0.0.0 (/checkout/compiler/rustc_interface)
 Documenting rustc_interface v0.0.0 (/checkout/compiler/rustc_interface)
[RUSTC-TIMING] rustc_interface test:false 1.639
 Documenting rustc_driver v0.0.0 (/checkout/compiler/rustc_driver)
Error: The operation was canceled.
```
I see nothing related to this PR's changes...

@bors retry

bors commented:
:hourglass: Testing commit e02b396d71d26f8899f797c49ee9c57254de1b48 with merge 80d5e84f2036b8c330d32e010f5ec86e96fe9d63...
<!-- homu: {"type":"BuildStarted","head_sha":"e02b396d71d26f8899f797c49ee9c57254de1b48","merge_sha":"80d5e84f2036b8c330d32e010f5ec86e96fe9d63"} -->

flip1995 commented on src/tools/clippy/clippy_lints/src/unnested_or_patterns.rs:
```diff
@@ -392,26 +402,42 @@ fn extend_with_tail_or(target: &mut Pat, tail_or: Vec<P<Pat>>) -> bool {
-    for pat in alternatives.drain_filter(|p| {
-        // Check if we should extract, but only if `idx >= start`.
+
+    // If `ThinVec` had the `drain_filter` method, this loop could be rewritten
```
Is changing things in Clippy to `ThinVec` necessary? Just looking at the Clippy changes seems that this just makes things worse, from a code style perspective. I would like if this could be avoided in Clippy and keep the old cleaner code.

nnethercote replied:
It can't be avoided. Clippy uses rustc's AST types. When those AST types change, Clippy has to change too.

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

nnethercote commented:
There is an infinite loop occurring in parallel-compiler builds, which can be replicated with this command:
```
x.py doc compiler/rustc_driver  --stage=1
```
The loop appears to involve `evaluate_trait_predicate_recursively` somehow. I suspect this is a latent bug in the parallel compiler that I am unluckily hitting :/

flip1995 commented on src/tools/clippy/clippy_lints/src/lib.rs:
```diff
@@ -43,6 +43,7 @@ extern crate rustc_session;
 extern crate rustc_span;
 extern crate rustc_target;
 extern crate rustc_trait_selection;
+extern crate thin_vec;
```
(un)educated guess here: You might have to specify the `thin-vec` dependency in [`rustc_workspace_hack`](https://github.com/rust-lang/rust/tree/master/src/tools/rustc-workspace-hack), but I'm not 100% sure that this will help. Using `extern crate` here is definitely something that we don't want to do here.

nnethercote replied:
I tried fiddling with `rustc_workspace_hack` but it didn't help. AIUI that's all about making sure that a consistent feature set is used for each crate, but that's not relevant here because rustc isn't using any `thin-vec` features.

> Is `thin_vec` shipped with `rustc-dev`?

I don't know what `rustc-dev` is. Some googling turned up https://rust-lang.github.io/rustup/concepts/components.html, which certainly sounds relevant. I found some mentions of `rustc-dev` in the rust repo, but I couldn't work out what crates are part of it, and how to add `thin_vec` to it. Any suggestions?

nnethercote replied:
It seems the `extern crate thin_vec` is needed because `ThinVec` is part of the interface for the AST, which clippy uses. From @cuviper: "If the crate is in any shared rustc API, then they have to use the same compilation of that crate."

rustdoc already has some `extern crate` declarations like this, and it gets another one in this PR for `thin_vec`. Seems like `clippy` is in the same situation.

nnethercote committed:
Upgrade `thin-vec` from 0.2.9 to 0.2.12.

Because 0.2.10 added supports for `ThinVec::splice`, and 0.2.12 is the
latest release.

bors commented:
:umbrella: The latest upstream changes (presumably #105554) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #105883) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #106156) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

the8472 commented:
> The loop appears to involve evaluate_trait_predicate_recursively somehow. I suspect this is a latent bug in the parallel compiler that I am unluckily hitting :/

That probably should be an issue and this should then be blocked on that.

the8472 commented:
@rustbot author

Nilstrieb commented:
Now that we have a second case of the bug, I created an issue: #106930

albertlarsan68 commented:
Tests have moved from `src/test` to `tests`, please [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing) on top of new `master`.
@rustbot label -A-bootstrap

nnethercote committed:
Use `ThinVec` in `ast::Block`.

nnethercote committed:
Use `ThinVec` in `ast::AngleBracketedArgs`.

nnethercote committed:
Use `ThinVec` in `ast::PatKind::Struct`.

nnethercote committed:
Use `ThinVec` in `ast::ExprKind::Match`.

nnethercote committed:
Use `ThinVec` in a few more AST types.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit fe5def512b1aff075262cc4dfa92445d6010642b with merge 0783cae66900d627183c1b16575e0ba0b56d90c1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fe5def512b1aff075262cc4dfa92445d6010642b","merge_sha":"0783cae66900d627183c1b16575e0ba0b56d90c1"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4040708063/jobs/6946604307) [(plain)](https://github.com/rust-lang/rust/commit/fe5def512b1aff075262cc4dfa92445d6010642b/checks/10965458498/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:ece894d15649f0f9d7884f915dc821f00bd0418b)
Complete job name: PR (x86_64-gnu-tools, false, 1, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-tools
---
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between f55b0022db8dccc6aa6bf3f650b562eaec0fdc54 and 8823e7c0c7402035a646cc03574f23f057e4c99a
Tool subtrees were updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
test test::verify_check_works_with_stdin ... ok
test test::verify_check_works ... ok
test test::configuration_snippet::configuration_snippet_tests ... ok

Mismatch at src/modules.rs:6:
 use rustc_ast::visit::Visitor;
 use rustc_span::symbol::{self, sym, Symbol};
 use rustc_span::Span;
 use thin_vec::ThinVec;
+use thiserror::Error;
 
 use crate::attr::MetaVisitor;
---
---- test::self_tests stdout ----
Ran 5 self tests.
thread 'test::self_tests' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: 1 self tests failed', src/tools/rustfmt/src/test/mod.rs:400:5


failures:
    test::self_tests
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4040709808/jobs/6948432476)
Build commit: 0783cae66900d627183c1b16575e0ba0b56d90c1 (`0783cae66900d627183c1b16575e0ba0b56d90c1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4040709808/jobs/6948432476"},"merge_sha":"0783cae66900d627183c1b16575e0ba0b56d90c1"} -->

rust-timer commented:
Queued 0783cae66900d627183c1b16575e0ba0b56d90c1 with parent f55b0022db8dccc6aa6bf3f650b562eaec0fdc54, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f55b0022db8dccc6aa6bf3f650b562eaec0fdc54&end=0783cae66900d627183c1b16575e0ba0b56d90c1).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (0783cae66900d627183c1b16575e0ba0b56d90c1): [comparison URL](https://perf.rust-lang.org/compare.html?start=f55b0022db8dccc6aa6bf3f650b562eaec0fdc54&end=0783cae66900d627183c1b16575e0ba0b56d90c1&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f55b0022db8dccc6aa6bf3f650b562eaec0fdc54&end=0783cae66900d627183c1b16575e0ba0b56d90c1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.4%, 0.7%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [0.3%, 6.2%]   | 20    |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.9%, -0.2%] | 27    |
| Improvements ✅ <br /> (secondary) | -0.5% | [-1.2%, -0.2%] | 22    |
| All ❌✅ (primary)                 | -0.3% | [-0.9%, 0.7%]  | 29    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f55b0022db8dccc6aa6bf3f650b562eaec0fdc54&end=0783cae66900d627183c1b16575e0ba0b56d90c1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [1.2%, 5.5%]   | 11    |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.5%, -0.5%] | 7     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-3.6%, -0.6%] | 10    |
| All ❌✅ (primary)                 | -1.0% | [-1.5%, -0.5%] | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f55b0022db8dccc6aa6bf3f650b562eaec0fdc54&end=0783cae66900d627183c1b16575e0ba0b56d90c1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.6%  | [4.6%, 4.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

lqd commented:
I'm trying #107121 with this PR, and it doesn't look like it fixes the issue happening when documenting `rustc_middle`.

the8472 commented:
ping @nnethercote it seems like @lqd's experiments show a potential fix that could unblock this PR

nnethercote commented:
Thanks for the info. I didn't see @lqd's update in the previous comment because GitHub doesn't send a notification email for comment updates :/

GuillaumeGomez committed:
Allow disabling of auto and blanket trait impls retrieval in rustdoc when in parallel_compiler mode.

m-ou-se committed:
Enable --cfg=parallel_compiler in rustdoc.

nnethercote committed:
Fix a test.

nnethercote commented:
It works!

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9c7570b4293f9eea593cdd8334d6ccba651f91c1 with merge 501502d1c672413aad29cf862cc45a53ac8c53c8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9c7570b4293f9eea593cdd8334d6ccba651f91c1","merge_sha":"501502d1c672413aad29cf862cc45a53ac8c53c8"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4228416453/jobs/7345496894)
Build commit: 501502d1c672413aad29cf862cc45a53ac8c53c8 (`501502d1c672413aad29cf862cc45a53ac8c53c8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4228416453/jobs/7345496894"},"merge_sha":"501502d1c672413aad29cf862cc45a53ac8c53c8"} -->

rust-timer commented:
Queued 501502d1c672413aad29cf862cc45a53ac8c53c8 with parent 8f55d6025fec5fb648948fbe4fb6dcb4184c67bc, future [comparison URL](https://perf.rust-lang.org/compare.html?start=8f55d6025fec5fb648948fbe4fb6dcb4184c67bc&end=501502d1c672413aad29cf862cc45a53ac8c53c8).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (501502d1c672413aad29cf862cc45a53ac8c53c8): [comparison URL](https://perf.rust-lang.org/compare.html?start=8f55d6025fec5fb648948fbe4fb6dcb4184c67bc&end=501502d1c672413aad29cf862cc45a53ac8c53c8&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=8f55d6025fec5fb648948fbe4fb6dcb4184c67bc&end=501502d1c672413aad29cf862cc45a53ac8c53c8&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.4%, 0.8%]   | 8     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.5%, -0.2%] | 76    |
| Improvements ✅ <br /> (secondary) | -0.8% | [-3.3%, -0.2%] | 42    |
| All ❌✅ (primary)                 | -0.4% | [-1.5%, 0.8%]  | 84    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8f55d6025fec5fb648948fbe4fb6dcb4184c67bc&end=501502d1c672413aad29cf862cc45a53ac8c53c8&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.0%  | [1.0%, 1.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [1.1%, 5.1%]   | 9     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.2%, -1.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-5.2%, -1.2%] | 9     |
| All ❌✅ (primary)                 | -0.4% | [-1.2%, 1.0%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8f55d6025fec5fb648948fbe4fb6dcb4184c67bc&end=501502d1c672413aad29cf862cc45a53ac8c53c8&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

nnethercote commented:
Perf results are still good.

@bors r=the8472

bors commented:
:pushpin: Commit 9c7570b4293f9eea593cdd8334d6ccba651f91c1 has been approved by `the8472`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=the8472 9c7570b4293f9eea593cdd8334d6ccba651f91c1 -->
<!-- homu: {"type":"Approved","sha":"9c7570b4293f9eea593cdd8334d6ccba651f91c1","approver":"the8472","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 9c7570b4293f9eea593cdd8334d6ccba651f91c1 with merge 3fee48c161a48b0c142d3998fff56faee96bd56c...
<!-- homu: {"type":"BuildStarted","head_sha":"9c7570b4293f9eea593cdd8334d6ccba651f91c1","merge_sha":"3fee48c161a48b0c142d3998fff56faee96bd56c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4230095786/jobs/7349748030)
Approved by: the8472
Pushing 3fee48c161a48b0c142d3998fff56faee96bd56c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"the8472","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4230095786/jobs/7349748030"},"merge_sha":"3fee48c161a48b0c142d3998fff56faee96bd56c"} -->

rust-timer commented:
Finished benchmarking commit (3fee48c161a48b0c142d3998fff56faee96bd56c): [comparison URL](https://perf.rust-lang.org/compare.html?start=f715e430aac0de131e2ad21804013ea405722a66&end=3fee48c161a48b0c142d3998fff56faee96bd56c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f715e430aac0de131e2ad21804013ea405722a66&end=3fee48c161a48b0c142d3998fff56faee96bd56c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.4%, 0.6%]   | 9     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.5%, -0.2%] | 80    |
| Improvements ✅ <br /> (secondary) | -0.8% | [-3.3%, -0.2%] | 44    |
| All ❌✅ (primary)                 | -0.4% | [-1.5%, 0.6%]  | 89    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f715e430aac0de131e2ad21804013ea405722a66&end=3fee48c161a48b0c142d3998fff56faee96bd56c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [2.2%, 4.1%]   | 8     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.5% | [-9.6%, -1.1%] | 14    |
| All ❌✅ (primary)                 | -0.6% | [-0.6%, -0.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f715e430aac0de131e2ad21804013ea405722a66&end=3fee48c161a48b0c142d3998fff56faee96bd56c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-2.8%, -2.8%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
